File: internal/service/usage_monitoring_types.go
package service

import "time"

type UsageMonitoringRecentRequest struct {
	Timestamp time.Time
	Failed    bool
}

type UsageMonitoringKPI struct {
	TotalRequests   int64
	SuccessRequests int64
	FailedRequests  int64
	TotalTokens     int64
	InputTokens     int64
	OutputTokens    int64
	CachedTokens    int64
	ReasoningTokens int64
	RPM             float64
	TPM             float64
	TotalCost       float64
	CostAvailable   bool
}

type UsageMonitoringModelDistributionItem struct {
	Model           string
	TotalRequests   int64
	SuccessCount    int64
	FailureCount    int64
	TotalTokens     int64
	InputTokens     int64
	OutputTokens    int64
	CachedTokens    int64
	ReasoningTokens int64
	SuccessRate     float64
}

type UsageMonitoringTrendPoint struct {
	Bucket          string
	Requests        int64
	Tokens          int64
	InputTokens     int64
	OutputTokens    int64
	CachedTokens    int64
	ReasoningTokens int64
	Cost            float64
}

type UsageMonitoringHourlyModelPoint struct {
	Hour   string
	Models []UsageMonitoringHourlyModelStat
}

type UsageMonitoringHourlyModelStat struct {
	Model        string
	Requests     int64
	Tokens       int64
	SuccessCount int64
	FailureCount int64
}

type UsageMonitoringChannelModelStat struct {
	Model           string
	Requests        int64
	Success         int64
	Failed          int64
	SuccessRate     float64
	TotalTokens     int64
	LastRequestTime *time.Time
	RecentRequests  []UsageMonitoringRecentRequest
}

type UsageMonitoringChannelStat struct {
	Source          string
	AuthIndex       string
	TotalRequests   int64
	SuccessRequests int64
	FailedRequests  int64
	TotalTokens     int64
	InputTokens     int64
	OutputTokens    int64
	CachedTokens    int64
	ReasoningTokens int64
	SuccessRate     float64
	LastRequestTime *time.Time
	RecentRequests  []UsageMonitoringRecentRequest
	Models          []UsageMonitoringChannelModelStat
}

type UsageMonitoringFailureModelStat struct {
	Model          string
	Success        int64
	Failure        int64
	Total          int64
	SuccessRate    float64
	LastTimestamp  *time.Time
	RecentRequests []UsageMonitoringRecentRequest
}

type UsageMonitoringFailureStat struct {
	Source       string
	AuthIndex    string
	FailedCount  int64
	LastFailTime *time.Time
	Models       []UsageMonitoringFailureModelStat
}

type UsageMonitoringRequestLog struct {
	ID              int64
	Timestamp       time.Time
	Model           string
	Source          string
	AuthIndex       string
	Failed          bool
	LatencyMS       int64
	InputTokens     int64
	OutputTokens    int64
	ReasoningTokens int64
	CachedTokens    int64
	TotalTokens     int64
}

type UsageMonitoringSnapshot struct {
	KPIs              UsageMonitoringKPI
	ModelDistribution []UsageMonitoringModelDistributionItem
	DailyTrend        []UsageMonitoringTrendPoint
	HourlyModelTrend  []UsageMonitoringHourlyModelPoint
	HourlyTokenTrend  []UsageMonitoringTrendPoint
	ChannelStats      []UsageMonitoringChannelStat
	FailureAnalysis   []UsageMonitoringFailureStat
	RequestLogs       []UsageMonitoringRequestLog
}

File: web/src/lib/usageMonitoringTypes.ts
import type { UsageEventTokens } from './types'

export interface UsageMonitoringRecentRequest {
  timestamp: string
  failed: boolean
}

export interface UsageMonitoringKPI {
  total_requests: number
  success_requests: number
  failed_requests: number
  total_tokens: number
  input_tokens: number
  output_tokens: number
  cached_tokens: number
  reasoning_tokens: number
  rpm: number
  tpm: number
  total_cost: number
  cost_available: boolean
}

export interface UsageMonitoringModelDistributionItem {
  model: string
  total_requests: number
  success_count: number
  failure_count: number
  total_tokens: number
  input_tokens: number
  output_tokens: number
  cached_tokens: number
  reasoning_tokens: number
  success_rate: number
}

export interface UsageMonitoringDailyTrendPoint {
  date: string
  requests: number
  tokens: number
  input_tokens: number
  output_tokens: number
  cached_tokens: number
  reasoning_tokens: number
  cost: number
}

export interface UsageMonitoringHourlyModelStat {
  model: string
  requests: number
  tokens: number
  success_count: number
  failure_count: number
}

export interface UsageMonitoringHourlyModelPoint {
  hour: string
  models: UsageMonitoringHourlyModelStat[]
}

export interface UsageMonitoringHourlyTokenPoint {
  hour: string
  input_tokens: number
  output_tokens: number
  cached_tokens: number
  reasoning_tokens: number
  total_tokens: number
  cost: number
}

export interface UsageMonitoringChannelModelStat {
  model: string
  requests: number
  success: number
  failed: number
  success_rate: number
  total_tokens: number
  last_request_time: string | null
  recent_requests: UsageMonitoringRecentRequest[]
}

export interface UsageMonitoringChannelStat {
  source: string
  source_type?: string
  source_key?: string
  total_requests: number
  success_requests: number
  failed_requests: number
  total_tokens: number
  input_tokens: number
  output_tokens: number
  cached_tokens: number
  reasoning_tokens: number
  success_rate: number
  last_request_time: string | null
  recent_requests: UsageMonitoringRecentRequest[]
  models: UsageMonitoringChannelModelStat[]
}

export interface UsageMonitoringFailureModelStat {
  model: string
  success: number
  failure: number
  total: number
  success_rate: number
  last_timestamp: string | null
  recent_requests: UsageMonitoringRecentRequest[]
}

export interface UsageMonitoringFailureStat {
  source: string
  source_type?: string
  source_key?: string
  failed_count: number
  last_fail_time: string | null
  models: UsageMonitoringFailureModelStat[]
}

export interface UsageMonitoringRequestLog {
  id?: number
  timestamp: string
  model: string
  reasoning_effort?: string
  source: string
  source_type?: string
  source_key?: string
  failed: boolean
  latency_ms: number
  tokens: UsageEventTokens
}

export interface UsageMonitoringResponse {
  kpis: UsageMonitoringKPI
  model_distribution: UsageMonitoringModelDistributionItem[]
  daily_trend: UsageMonitoringDailyTrendPoint[]
  hourly_model_trend: UsageMonitoringHourlyModelPoint[]
  hourly_token_trend: UsageMonitoringHourlyTokenPoint[]
  channel_stats: UsageMonitoringChannelStat[]
  failure_analysis: UsageMonitoringFailureStat[]
  request_logs: UsageMonitoringRequestLog[]
  timezone: string
  range_start?: string
  range_end?: string
}
